### PR TITLE
Add parent tracking to visitor

### PR DESCRIPTION
**Pull request by s-coimbra21**
Add parent tracking using a stack to track the path from the current node being visited to the tree root. I'm using a `ParentNode` enum with two variants, one for `ExpressionNode` and another for `StatementNode` which covers all possible visitable nodes that may have children, the variants store a reference to the actual node so I had to change the lifetimes to &'ast explicitly on the `Visitable` trait, perhaps there is a better way.

### CODE REVIEW

**Comment by maciejhirsz** [comment]
Will review in depth when I can, but just glossing over the code, it looks really good!

File: ratel-visitor/src/statement.rs [hunk 1]
```diff
@@ -1,102 +1,121 @@
...
         use self::Statement::*;
 
         match self.item {
-            Empty => {},
-            Expression(ref expression)   => {
-                expression.traverse(visitor, ctx);
+            Empty => { return; },
+            Expression(ref expression) => {
                 visitor.on_expression_statement(expression, self, ctx);
+                visitor.push_parent(ParentNode::from(self), ctx);
```
[the remark this answers is not in the document]
Comment (s-coimbra21) [reply ratel-visitor/src/statement.rs]: I don't think it's possible to adapt the macro to allow the type argument is it? I'm thinking I will just implement it separately inside each visitor on `build.rs`. Though I don't see the compiler generating code for the `ParentNode::from` lines if `push_parent` doesn't use it, it doesn't even have to be a noop.

File: .gitignore [hunk 2]
```diff
@@ -2,3 +2,4 @@
 /Cargo.lock
 */target/
 */Cargo.lock
+rls/
```
Comment (cmtt) [inline .gitignore]: Please ensure that the file is saved with a trailing newline.

File: ratel-visitor/src/statement.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (maciejhirsz) [reply ratel-visitor/src/statement.rs]: Ok, I slept on it a bit. If there are no performance concerns I think it's fine. The code of the library could be cleaner, but on the flip side the `Visitor` would have to call `.into()` all the time in `impl`s, which would make the API more annoying to use, so I reckon it's fine as is.

**Pull request merged**